kore/src/Kore/ASTVerifier/PatternVerifier.hs

```diff
@@ -571,21 +567,46 @@ verifyPatternsWithSorts getChildAttributes sorts operands = do
     actualOperandCount = length operands
 
 verifyApplyAlias
-    ::  (child -> Attribute.Pattern Variable)
-    ->  Application SymbolOrAlias (PatternVerifier child)
+    ::  (Verified.Pattern -> Attribute.Pattern Variable)
+    ->  Application SymbolOrAlias (PatternVerifier Verified.Pattern)
     ->  MaybeT PatternVerifier
             (CofreeF
                 (Application (Internal.Alias (TermLike Variable)))
                 (Attribute.Pattern Variable)
-                child
+                Verified.Pattern
             )
 verifyApplyAlias getChildAttributes application =
     lookupAlias symbolOrAlias >>= \alias -> Trans.lift $ do
     let verified = application { applicationSymbolOrAlias = alias }
         sorts = Internal.aliasSorts alias
+    leftVariables <- getLeftVariables (Internal.aliasConstructor alias)
+    Foldable.traverse_ ensureChildIsDeclaredVarType $ zip leftVariables children
     verifyApplicationChildren getChildAttributes verified sorts
   where
-    Application { applicationSymbolOrAlias = symbolOrAlias } = application
+    Application
+        { applicationSymbolOrAlias = symbolOrAlias
+        , applicationChildren      = children
+        } = application
+
+    getLeftVariables :: Id -> PatternVerifier [UnifiedVariable Variable]
+    getLeftVariables aliasId = do
+        indexedModule <- Reader.asks indexedModule
+        alias <- resolveAlias indexedModule aliasId
+        pure . applicationChildren . sentenceAliasLeftPattern . snd $ alias
+
+    -- If an alias was defined using an element variable, it can only take an
+    -- argument that is an element variable.
+    -- If it was defined using a set variable, we can use it with any argument.
+    -- Otherwise, it is a verification error.
+    ensureChildIsDeclaredVarType
+        :: (UnifiedVariable Variable, PatternVerifier Verified.Pattern)
+        -> PatternVerifier ()
+    ensureChildIsDeclaredVarType (var, mpat) = do
+        pat <- mpat
+        case (var, pat) of
+            (ElemVar _, Internal.ElemVar_ _) -> pure ()
+            (SetVar  _, _) -> pure ()
+            _ -> koreFailText "Child is not the expected type of variable."
```

Comment:
This error can be more specific: the alias was declared with an element variable, so its argument in that position must be an element variable. We should also include the location of `pat`.

Reply:
Not quite sure how to add the location of `pat`, since it's just a `TermLike`. Am I missing something? I changed the text though.